crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -217,18 +213,57 @@
 //! snapshot (which has `x = 3` as the only live binding). The result of this merge is that we now
 //! have two live bindings of `x`: `x = 3` and `x = 4`.
 //!
+//! Another piece of information that the `UseDefMap` needs to provide are visibility constraints.
+//! These are similar to the narrowing constraints, but apply to bindings and declarations within a
+//! control flow path. Consider the following example:
+//! ```py
+//! x = 1
+//! if test:
+//!     x = 2
+//!     y = "y"
+//! ```
+//! In principle, there are two possible control flow paths here. However, if we can statically
+//! infer `test` to be of type `Literal[True]` or `Literal[False]`, we can rule out one of the
+//! possible paths. To support this feature, we record a visibility constraint of `test` to all
+//! live bindings and declarations *after* visiting the body of the `if` statement. And we record
+//! a negative visibility constraint `~test` to all live bindings/declarations in the (implicit)
+//! `else` branch. For the example above, we would record the following visibility constraints
+//! (adding the implicit "unbound" definitions for clarity):
+//! ```py
+//! x = <unbound>  # not live, shadowed by `x = 1`
+//! y = <unbound>  # visibility constraint: ~test
+//!
+//! x = 1  # visibility constraint: ~test
+//! if test:
+//!     x = 2  # visibility constraint: test
+//!     y = "y"  # visibility constraint: test
+//! ```
+//! When we encounter a use of `x` after this `if` statement, we would record two live bindings: `x
+//! = 1` with a constraint of `~test`, and `x = 2` with a constraint of `test`. In type inference,
+//! when we iterate over all live bindings, we can evaluate these constraints to determine if a
+//! particular binding is actually visible. For example, if `test` is `Literal[True]`, we only see
+//! the `x = 2` binding. If `test` is `Literal[False]`, we only see the `x = 1` binding. And if the
+//! type of `test` is `bool`, we can see both bindings.
```

Comment:
> Python automatically converts an expression of any type to bool in a boolean context

Right. I was wondering whether or not I should bring it up and decided to keep it simple, because I didn't want to talk too much about types in the use-def-map module. But it's better to be precise, rather than concise. Changed now.